framework/src/org/checkerframework/common/fieldtypeinference/FieldTypeInferenceAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,89 @@
+package org.checkerframework.common.fieldtypeinference;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import org.checkerframework.common.basetype.BaseAnnotatedTypeFactory;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.dataflow.analysis.FlowExpressions.Receiver;
+import org.checkerframework.framework.flow.CFAbstractAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.util.AnnotatedTypes;
+
+import com.sun.source.tree.CompilationUnitTree;
+
+/**
+ * The ATF of a Checker must be a subtype of this class for it to perform
+ * intra-class inference for private fields.
+ *
+ * @author pbsf
+ *
+ */
+public class FieldTypeInferenceAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    /** Maps a Receiver representing a private field to an ATM that contains the
+    LUB of all AMs they have been assigned to. **/
+    private final Map<Receiver, AnnotatedTypeMirror> assignedPrivateFields =
+            new HashMap<Receiver, AnnotatedTypeMirror>();
+
+    public FieldTypeInferenceAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+    }
+
+    @Override
+    public void setRoot(CompilationUnitTree root) {
+        super.setRoot(root);
+        assignedPrivateFields.clear();
+    }
+
+    @Override
+    public CFTransfer createFlowTransferFunction(
+            CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+        return new FieldTypeInferenceTransfer(analysis);
+    }
+
+    /**
+     * Updates the ATM mapped by the parameter <tt>r</tt> in the map
+     * <tt>assignedPrivateFields</tt> to be the LUB between the parameter
+     * <tt>rhsATM</tt> and the current ATM mapped by <tt>r</tt> in this map.
+     * <p>
+     * If <tt>r</tt> is not a key in the map, it is added to the map
+     * mapping to <tt>rhsATM</tt>.
+     * @param r
+     *      A receiver representing a private field.
+     * @param fieldDeclType
+     *      The declared type of the private field.
+     * @param rhsATM
+     *      An ATM assigned to the private field in <tt>r</tt>.
+     */
+    public void addAssignedField(Receiver r, AnnotatedTypeMirror fieldDeclType,
```

Comment:
This side-effects fieldDeclType.  That was a surprise to me.  I see it's commented at the caller, but it's still surprising and isn't documented here.  It also seems inefficient:  why do the copy at every call when the body might not need the copy in most circumstances?  I think it would be clearer for this method to do the copy if needed, and for the client not do do any copying.  The specification of the routine would be simpler if it doesn't do the side effect, too.